[JIT] Handle aliasing cases where two arguments alias in SchemaCheckMode

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79991
* #79904
* __->__ #79814

- Modified SchemaCheckMode class to precompute a set of aliasing arguments for each input argument so that it can ignore cases where outputs mutate/alias due to inputs aliasing before an op occurs
- Tested through a case such that an input is mutated and is aliasing but the schema does not predict that it should
- Also tested through correct functionality of all other test cases

Commit:
[JIT] Handle aliasing cases where two arguments alias in SchemaCheckMode

[ghstack-poisoned]

Commit:
Update on "[JIT] Handle aliasing cases where two arguments alias in SchemaCheckMode"


- Modified SchemaCheckMode class to precompute a set of aliasing arguments for each input argument so that it can ignore cases where outputs mutate/alias due to inputs aliasing before an op occurs
- Tested through a case such that an input is mutated and is aliasing but the schema does not predict that it should
- Also tested through correct functionality of all other test cases

[ghstack-poisoned]

Commit:
Update on "[JIT] Handle aliasing cases where two arguments alias in SchemaCheckMode"


- Modified SchemaCheckMode class to precompute a set of aliasing arguments for each input argument so that it can ignore cases where outputs mutate/alias due to inputs aliasing before an op occurs
- Tested through a case such that an input is mutated and is aliasing but the schema does not predict that it should
- Also tested through correct functionality of all other test cases

[ghstack-poisoned]

davidberard98 commented on torch/testing/_internal/schema_check_mode.py:
```diff
@@ -56,22 +72,32 @@ def unwrap(e):
         cloned_arguments = dict(zip(arguments.keys(), clone_inputs(arguments.values())))
         out = func(*args, **kwargs)
 
+        arg_alias_pairs_map = {arg.name : [arg] for arg in func._schema.arguments}
```
@goldenxuett actually I meant a comment describing why we need this in the first place (e.g. provide context about why aliasing inputs will affect the expected outputs that are aliasing)

davidberard98 reviewed (approved):
Sorry - should have clarified this on the first review but could you modify the comment a bit?

Otherwise looks good - feel free to update the comment and then merge as soon as the tests are green

Commit:
Update on "[JIT] Handle aliasing cases where two arguments alias in SchemaCheckMode"


- Modified SchemaCheckMode class to precompute a set of aliasing arguments for each input argument so that it can ignore cases where outputs mutate/alias due to inputs aliasing before an op occurs
- Tested through a case such that an input is mutated and is aliasing but the schema does not predict that it should
- Also tested through correct functionality of all other test cases

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
